kedro/config/abstract_config.py

```diff
@@ -1,11 +1,11 @@
 """This module provides ``kedro.abstract_config`` with the baseline
 class model for a `ConfigLoader` implementation.
 """
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from typing import Any, Dict
 
 
-class AbstractConfigLoader(ABC):
+class AbstractConfigLoader(dict):
```

Comment:
I think it would be better to inherit from `UserDict` instead of the `dict`, there are some traps with inheriting from the Python standard `dict` class.

It would also be more consistent, see `_ProjectLogging` that we are already using `UserDict`.
https://github.com/kedro-org/kedro/blob/4db0ae3be611f60f3c98da2501364d5549c784f2/kedro/framework/project/__init__.py#L199-L208

Fluent Python has a detail explanation for it.

See these two relevant chapter
* https://learning.oreilly.com/library/view/fluent-python-2nd/9781492056348/ch14.html#subclass_builtin_woes
* https://learning.oreilly.com/library/view/fluent-python-2nd/9781492056348/ch03.html#missing_method

Reply:
I don't remember this 100% anymore, but I think initially the plan was to have [`AbstractConfigLoader` inherit from `MutableMapping`](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/#Abstract_base_classes:_they%E2%80%99ll_help_you_quack_like_a_duck) but then it ended up being a [lot of overwriting of stuff we don't need.](https://github.com/kedro-org/kedro/commit/6dc6eb151bc9112752d6fe669912c689a1921732) And then the decision was made to inherit from `dict` directly. This was all over Slack and not very coherent. 

Thinking about this again now, I feel like you might be right and we don't actually need to inherit from `dict` at all since we're overwriting `get` and `__getitem__` anyway.